File: acoustic_sensor_recording/win/src/Beehive_Multimodal_Acoustic_Recording--Event-Only.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Code definition: 

# Using sounddevice and soundfile libraries, record audio from a device ID and save it to a FLAC file.
# Input audio from a device ID at a defineable sample rate, bit depth, and channel count. 
# Write incoming audio into a circular buffer that is of a definable length. 
# Monitor the incoming audio for levels above a definable threshold for a defineable duration and set a flag when conditions are met. 
# Note the position in the buffer of the event and then continue to record audio until a definable time period after the start of the event. 
# Note the position in the buffer of the end of the time period after the start of the event.
# Continue recording audio into the circular buffer while saving the audio to a FLAC file.
# Save audio in the circular buffer from the start of a defineable time period before the event to the end of the defineable time period after the event.
# Reset the audio threshold level flag and event_start_time after saving audio.


import sounddevice as sd
import soundfile as sf
from datetime import datetime
import time
import os
import io
import threading
import numpy as np
from scipy.signal import resample_poly
##from pydub import AudioSegment


THRESHOLD = 27000            # audio level threshold to be considered an event
BUFFER_SECONDS = 400        # seconds of a circular buffer
SAMPLE_RATE = 192000         # Audio sample rate
DEVICE_IN = 1               # Device ID of input device
DEVICE_OUT = 3              # Device ID of output device
BIT_DEPTH = 16              # Audio bit depth
CHANNELS = 2                # Number of channels
##OUTPUT_DIRECTORY = "."      # for debugging
OUTPUT_DIRECTORY = "D:/OneDrive/data/Zeev/recordings"
FORMAT = 'FLAC'             # 'WAV' or 'FLAC'INTERVAL = 0 # seconds between recordings

# event recording
SAVE_BEFORE_EVENT = 30   # seconds to save before the event
SAVE_AFTER_EVENT = 30    # seconds to save after the event

# init event variables
event_start_index = None
event_save_thread = None
detected_level = None

_dtype = None
_subtype = None


# Op Mode & ID =====================================================================================
#MODE = "event"             # event only
MODE = "combo"             # period recording with event detection
LOCATION_ID = "Zeev-Berkeley"
HIVE_ID = "Z1"
# ==================================================================================================

### startup housekeeping ###
def initialization():
    global buffer, buffer_index, _dtype, buffer_size, _subtype

    # Check on parms
    if (SAVE_BEFORE_EVENT + SAVE_AFTER_EVENT) * 1.2 > BUFFER_SECONDS:
        print("The buffer is not large enough to hold the maximum amount of audio that can be saved.")
        print("Reduce SAVE_DURATION_BEFORE and/or SAVE_DURATION_AFTER or increase the size of the circular buffer 'BUFFER_SECONDS'")
        quit(-1)

    if (PERIOD) * 1.1 > BUFFER_SECONDS:
        print("The buffer is not large enough to hold the maximum amount of audio that can be saved.")
        print("Reduce PERIOD or increase the size of the circular buffer 'BUFFER_SECONDS'")
        quit(-1)

    # Check on input device parms or if input device even exits
    try:
        device_info = sd.query_devices(DEVICE_IN)  
        device_CH = device_info['max_input_channels'] 
        if device_CH < CHANNELS:
            print(f"The device only has {device_CH} channel(s) but require {CHANNELS} channels.")
            quit(-1)
        ##device_SR = device_info['default_samplerate'] 
        ##if device_SR != SAMPLE_RATE:
        ##    print(f"The device sample rate {device_SR} is not equal to the required 'SAMPLE_RATE' of {SAMPLE_RATE}")
        ##    quit(-1)
    except Exception as e:
        print(f"An error occurred while attempting to access the input device: {e}")
        print("These are the available devices: \n", sd.query_devices())
        quit(-1)

    # Create the output directory if it doesn't exist
    try:
        os.makedirs(OUTPUT_DIRECTORY, exist_ok=True)
    except Exception as e:
        print(f"An error occurred while trying to make or find output directory: {e}")
        quit(-1)

    # translate human to machine
    if BIT_DEPTH == 16:
        _dtype = 'int16'
        _subtype = 'PCM_16'
    elif BIT_DEPTH == 24:
        _dtype = 'int24'
        _subtype = 'PCM_24'
    elif BIT_DEPTH == 32:
        _dtype = 'int32' 
        _subtype = 'PCM_32'
    else:
        print("The bit depth is not supported: ", BIT_DEPTH)
        quit(-1)

    # audio buffers and variables
    buffer_size = int(BUFFER_SECONDS * SAMPLE_RATE)
    buffer = np.zeros((buffer_size, CHANNELS), dtype=_dtype)
    buffer_index = 0

#
# event recording functions
#
def save_audio_around_event():
    time.sleep(SAVE_AFTER_EVENT)
    save_event_audio()


def save_event_audio():
    global buffer, event_start_index, event_save_thread, detected_level

    if event_start_index is None:  # if this has been reset already, don't try to save
        return

    save_start_index = (event_start_index - SAVE_BEFORE_EVENT * SAMPLE_RATE) % buffer_size
    save_end_index = (event_start_index + SAVE_AFTER_EVENT * SAMPLE_RATE) % buffer_size

    # saving from a circular buffer so segments aren't necessarily contiguous
    if save_end_index > save_start_index:   # is contiguous
        audio_data = buffer[save_start_index:save_end_index]
    else:                                   # ain't contiguous
        audio_data = np.concatenate((buffer[save_start_index:], buffer[:save_end_index]))

    timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
    threshold_tag = int(THRESHOLD/1000)
    output_filename = f"{timestamp}_event_{detected_level}_{SAVE_BEFORE_EVENT}_{SAVE_AFTER_EVENT}_{LOCATION_ID}_{HIVE_ID}.{FORMAT.lower()}"
    full_path_name = os.path.join(OUTPUT_DIRECTORY, output_filename)
    sf.write(full_path_name, audio_data, SAMPLE_RATE, format=FORMAT, subtype=_subtype)

    print(f"Saved evemt audio to {full_path_name}, audio threshold level: {detected_level}, duration: {audio_data.shape[0] / SAMPLE_RATE} seconds")

    event_save_thread = None
    event_start_index = None


def check_level(audio_data, index):
    global event_start_index, event_save_thread, detected_level

    audio_level = np.max(np.abs(audio_data))
    if (audio_level > THRESHOLD) and event_start_index is None:
        print("event detected at:", datetime.now(), "audio level:", audio_level)
        detected_level = audio_level
        event_start_index = index
        event_save_thread = threading.Thread(target=save_audio_around_event)
        event_save_thread.start()
#
# audio stream callback function
#
def callback(indata, frames, time, status):
    global buffer, buffer_index

    if status:
        print("Callback status:", status)

    data_len = len(indata)
    # managing a circular buffer
    if buffer_index + data_len <= buffer_size:
        buffer[buffer_index:buffer_index + data_len] = indata
    else:
        overflow = (buffer_index + data_len) - buffer_size
        buffer[buffer_index:] = indata[:-overflow]
        buffer[:overflow] = indata[-overflow:]

    check_level(indata, buffer_index)   # trigger saving audio if above threshold

    buffer_index = (buffer_index + data_len) % buffer_size


def audio_stream():
    global buffer, buffer_index, _dtype

    stream = sd.InputStream(device=DEVICE_IN, channels=CHANNELS, samplerate=SAMPLE_RATE, dtype=_dtype, callback=callback)
    with stream:
        print("Start recording...")
        while stream.active:
            pass

###########################
########## MAIN ###########
###########################

if __name__ == "__main__":

    initialization()
    print("Acoustic Signal Capture")
    print(f"Sample Rate: {SAMPLE_RATE}; File Format: {FORMAT}; Channels: {CHANNELS}")
    try:
        if MODE == 'orig_period':
            print("Starting audio stream in original period-only recording mode")
            period_segment_recording()
        elif MODE == 'period':
            print("Starting audio stream in period-only mode")
            audio_stream()
        elif MODE == 'event':
            print("Starting audio stream in event detect-only mode")
            audio_stream()
        elif MODE == 'combo':
            print("Starting audio capture in periodic mode w/event capture")
            audio_stream()
        else:
            print("MODE not recognized")
            quit(-1)
    except KeyboardInterrupt:
        print('\nRecording process stopped by user.')
    except Exception as e:
        print(f"An error occurred while attempting to execute this script: {e}")
        quit(-1)         # quit with error

File: acoustic_sensor_recording/win/src/Beehive_Multimodal_Acoustic_Recording--Event-Only.cpp
#!/usr/bin/env c++

#include <iostream>
#include <chrono>
#include <thread>
#include <boost/circular_buffer.hpp>
#include <boost/thread.hpp>
#include <portaudio.h>
#include <sndfile.h>

#define THRESHOLD 27000
#define BUFFER_SECONDS 400
#define SAMPLE_RATE 192000
#define CHANNELS 2
#define BIT_DEPTH 16

boost::circular_buffer<double> buffer(BUFFER_SECONDS * SAMPLE_RATE * CHANNELS);
unsigned long buffer_index = 0;

std::string MODE = "combo";

int save_before_event = 30;
int save_after_event = 30;
int detected_level = 0;
int event_start_index = -1;
boost::thread* event_save_thread = nullptr;

int callback(const void* inputBuffer, void* outputBuffer, unsigned long framesPerBuffer,
             const PaStreamCallbackTimeInfo* timeInfo, PaStreamCallbackFlags statusFlags, void* userData) {
    auto in = (float*)inputBuffer;
    for (unsigned long i = 0; i < framesPerBuffer; i++) {
        for (int j = 0; j < CHANNELS; j++) {
            buffer.push_back(in[i * CHANNELS + j]);
        }
    }

    // Check level and start event save thread if necessary
    for (unsigned long i = 0; i < framesPerBuffer; i++) {
        if (std::abs(in[i]) > THRESHOLD && event_start_index == -1) {
            event_start_index = buffer_index;
            event_save_thread = new boost::thread(save_audio_around_event);
        }
    }
    buffer_index += framesPerBuffer;
    return paContinue;
}

void save_audio_around_event() {
    std::this_thread::sleep_for(std::chrono::seconds(save_after_event));
    save_event_audio();
}

void save_event_audio() {
    // Saving from a circular buffer
    // Not implemented in this example
}

int main() {
    PaStream* stream;
    Pa_Initialize();
    Pa_OpenDefaultStream(&stream, CHANNELS, 0, paFloat32, SAMPLE_RATE, 256, callback, nullptr);
    Pa_StartStream(stream);
    while (true) {
        Pa_Sleep(1000);
    }
    Pa_StopStream(stream);
    Pa_Terminate();
    return 0;
}


/*
The following is a conversion of your Python code to C++. It must be noted that C++ lacks many of the high-level features provided by Python, particularly in regards to audio processing and threading. Therefore, you'll need to use external libraries to accomplish the same functionality in C++. In this conversion, I'll be using the following libraries:

PortAudio for audio I/O
libsndfile for audio file handling
boost for threading and circular buffer
The provided code should be treated as a guideline, as it's not fully functional. Some of the more complex features in your Python script are beyond the scope of this conversion and would require extensive work and possible restructuring of the code to get working in C++. However, this should give you a solid starting point.

A couple of important notes:

You will need to properly setup and compile the code with the mentioned libraries.
As of my knowledge cutoff in September 2021, there is no library in C++ that natively supports 24-bit PCM, similar to Python's sounddevice library. Therefore, handling of 24-bit audio might be more complex and it's not shown in the code above.
Threading is more complex in C++ compared to Python. The code above simply creates a new thread to run the save_audio_around_event function and does not manage the lifetime of the thread. In a real program, you would want to properly handle thread lifetimes, which could include joining or detaching threads at appropriate times.
This code does not handle any file operations. You would need to implement the save_event_audio function yourself, which should write the contents of the buffer to a file. You could use the libsndfile library for this.
The callback function in the code above is a minimal implementation that simply fills a buffer with the input audio data. It does not handle overflows like the Python code does.
This code does not check the capabilities of the audio device, like the Python code does.
Error handling is minimal in the code above.
As the original Python code did not provide a full working example (missing imports, functions), the conversion might not include all functionality that you require.
As C++ and Python are very different languages, this conversion may not perform exactly the same as the original Python script. Please review the C++ code carefully and adapt as necessary for your specific application.
Please note that audio processing in C++ can be complex and require a solid understanding of both C++ and digital audio. If you're not comfortable with C++, you might want to consider using a higher-level language or a specialized audio processing framework or tool.

*/

// version of 7/21/23

Python and C++ are quite different in nature, and some Python modules such as `sounddevice`, `soundfile`, and `numpy` do not have direct equivalents in C++. However, C++ has libraries like PortAudio for audio I/O and libsndfile for reading/writing audio files, which can be used to perform similar operations. Here's a rough idea of how the script might be translated into C++.

Please note that this code is a simplified and partial translation of the Python script you provided. It does not include all the functionalities in the original Python script. For example, it does not handle FLAC files, as support for FLAC is not included in libsndfile by default. Also, threading, exceptions, and some other parts are omitted.

```cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <chrono>
#include <thread>
#include <sndfile.hh>
#include <portaudio.h>

#define SAMPLE_RATE 192000
#define CHANNELS 2
#define FRAMES_PER_BUFFER 1024
#define THRESHOLD 27000
#define SAVE_BEFORE_EVENT 30
#define SAVE_AFTER_EVENT 30

typedef struct {
    double buffer[FRAMES_PER_BUFFER * CHANNELS];
    int frameIndex;
} paTestData;

std::vector<double> audioBuffer;
int eventStartIndex = -1;
std::string OUTPUT_DIRECTORY = "D:/OneDrive/data/Zeev/recordings";

static int recordCallback(const void *inputBuffer, void *outputBuffer, unsigned long framesPerBuffer,
                          const PaStreamCallbackTimeInfo* timeInfo, PaStreamCallbackFlags statusFlags, void *userData) {
    paTestData *data = (paTestData*)userData;
    const double *rptr = (const double*)inputBuffer;
    double *wptr = &data->buffer[data->frameIndex * CHANNELS];
    long framesToCalc;
    long i;
    int finished;

    if(data->frameIndex < FRAMES_PER_BUFFER) {
        framesToCalc = FRAMES_PER_BUFFER - data->frameIndex;
        finished = paContinue;
    } else {
        framesToCalc = framesPerBuffer;
        finished = paComplete;
    }

    if(inputBuffer == NULL) {
        for(i=0; i<framesToCalc; i++) {
            *wptr++ = 0.0;
            if(CHANNELS == 2) *wptr++ = 0.0;
        }
    } else {
        for(i=0; i<framesToCalc; i++) {
            audioBuffer.push_back(*rptr);
            *wptr++ = *rptr++;
            if(CHANNELS == 2) {
                audioBuffer.push_back(*rptr);
                *wptr++ = *rptr++;
            }
        }
    }

    data->frameIndex += framesToCalc;
    if(finished) {
        if (eventStartIndex != -1) {
            saveEventAudio();
        }
    }
    return finished;
}

void saveEventAudio() {
    // Logic to save audio from eventStartIndex - SAVE_BEFORE_EVENT * SAMPLE_RATE to eventStartIndex + SAVE_AFTER_EVENT * SAMPLE_RATE
    // Reset eventStartIndex to -1
}

void checkLevel() {
    for (int i = 0; i < audioBuffer.size(); i++) {
        if (std::abs(audioBuffer[i]) > THRESHOLD && eventStartIndex == -1) {
            std::cout << "Event detected at: " << std::chrono::system_clock::now().time_since_epoch().count() << "audio level: " << audioBuffer[i] << "\n";
            eventStartIndex = i;
            std::this_thread::sleep_for(std::chrono::seconds(SAVE_AFTER_EVENT));
        }
    }
}

void startRecording() {
    PaStreamParameters inputParameters;
    PaStream *stream;
    paTestData data;
    PaError err = paNoError;

    err = Pa_Initialize();
    if(err != paNoError) std::cout << "Error in Pa_Initialize: " << Pa_GetErrorText(err) << "\n";

    inputParameters.device = Pa_GetDefaultInputDevice();
    if(inputParameters.device == paNoDevice) std::cout << "Error: No default input device.\n";

    inputParameters.channelCount = CHANNELS;
    inputParameters.sampleFormat = paFloat32;
    inputParameters.suggestedLatency = Pa_GetDeviceInfo(inputParameters.device)->defaultLowInputLatency;
    inputParameters.hostApiSpecificStreamInfo = NULL;

    err = Pa_OpenStream(&stream, &inputParameters, NULL, SAMPLE_RATE, FRAMES_PER_BUFFER, paClipOff, recordCallback, &data);
    if(err != paNoError) std::cout << "Error in Pa_OpenStream: " << Pa_GetErrorText(err) << "\n";

    err = Pa_StartStream(stream);
    if(err != paNoError) std::cout << "Error in Pa_StartStream: " << Pa_GetErrorText(err) << "\n";

    while((err = Pa_IsStreamActive(stream)) == 1) {
        checkLevel();
    }

    err = Pa_CloseStream(stream);
    if(err != paNoError) std::cout << "Error in Pa_CloseStream: " << Pa_GetErrorText(err) << "\n";

    Pa_Terminate();
}

int main() {
    startRecording();
    return 0;
}
```
/* 
This script initializes a PortAudio stream, then begins recording audio data. As data is recorded, it is added to a buffer, and the `checkLevel` function checks if the audio level exceeds the threshold. If it does, it triggers the save event audio function after a delay.

Please note that you'll need to link against PortAudio and libsndfile when compiling this C++ code. This can be done by adding `-lportaudio -lsndfile` to your `g++` command.

Also, remember that the C++ standard library does not have built-in support for working with directories, so you will need to use the appropriate OS-specific API or a library like Boost.Filesystem to create directories. 

Please consider this as a starting point and you will need to modify this to meet your exact requirements.
*/

// version of 7/21/23 using a circular buffer

/* Here's a basic example of how you might implement this in C++ using the PortAudio and libsndfile libraries. This code sets up a simple audio stream with PortAudio, saves incoming audio data into a circular buffer, and provides a function to save a portion of that buffer to a WAV file using libsndfile.
*/

#!/usr/bin/env c++

#include <iostream>
#include <vector>
#include <portaudio.h>
#include <sndfile.hh>

#define SAMPLE_RATE 192000
#define CHANNELS 2
#define FRAMES_PER_BUFFER 1024
#define BUFFER_SIZE 1000000

std::vector<short> circularBuffer(BUFFER_SIZE);

int bufferIndex = 0;

int recordCallback(const void *inputBuffer, void *outputBuffer, unsigned long framesPerBuffer,
                   const PaStreamCallbackTimeInfo* timeInfo, PaStreamCallbackFlags statusFlags, void *userData) {
    const short *rptr = (const short*)inputBuffer;
    for(int i = 0; i < framesPerBuffer * CHANNELS; i++) {
        circularBuffer[bufferIndex] = *rptr++;
        bufferIndex = (bufferIndex + 1) % BUFFER_SIZE;
    }
    return paContinue;
}

void saveToFile(int start, int end, const char* filename) {
    SF_INFO sfinfo;
    sfinfo.channels = CHANNELS;
    sfinfo.samplerate = SAMPLE_RATE;
    sfinfo.format = SF_FORMAT_WAV | SF_FORMAT_PCM_16;

    SndfileHandle outfile(filename, SFM_WRITE, sfinfo.format, sfinfo.channels, sfinfo.samplerate);

    if(end >= start) {
        outfile.write(&circularBuffer[start], end - start);
    } else {
        outfile.write(&circularBuffer[start], BUFFER_SIZE - start);
        outfile.write(&circularBuffer[0], end);
    }
}

int main() {
    PaStreamParameters inputParameters;
    PaStream *stream;
    PaError err = paNoError;

    err = Pa_Initialize();
    if(err != paNoError) std::cout << "Error in Pa_Initialize: " << Pa_GetErrorText(err) << "\n";

    inputParameters.device = Pa_GetDefaultInputDevice();
    if(inputParameters.device == paNoDevice) std::cout << "Error: No default input device.\n";

    inputParameters.channelCount = CHANNELS;
    inputParameters.sampleFormat = paInt16;
    inputParameters.suggestedLatency = Pa_GetDeviceInfo(inputParameters.device)->defaultLowInputLatency;
    inputParameters.hostApiSpecificStreamInfo = NULL;

    err = Pa_OpenStream(&stream, &inputParameters, NULL, SAMPLE_RATE, FRAMES_PER_BUFFER, paClipOff, recordCallback, NULL);
    if(err != paNoError) std::cout << "Error in Pa_OpenStream: " << Pa_GetErrorText(err) << "\n";

    err = Pa_StartStream(stream);
    if(err != paNoError) std::cout << "Error in Pa_StartStream: " << Pa_GetErrorText(err) << "\n";

    Pa_Sleep(5000);  // Record for 5 seconds.

    err = Pa_StopStream(stream);
    if(err != paNoError) std::cout << "Error in Pa_StopStream: " << Pa_GetErrorText(err) << "\n";

    saveToFile(0, bufferIndex, "output.wav");

    err = Pa_CloseStream(stream);
    if(err != paNoError) std::cout << "Error in Pa_CloseStream: " << Pa_GetErrorText(err) << "\n";

    Pa_Terminate();

    return 0;
}

/*
This program will record audio for 5 seconds, then save the audio to a WAV file. The audio is stored in a circular buffer implemented as a `std::vector`. When the buffer becomes full, the program will start overwriting the oldest data.
The `saveToFile` function saves a specified portion of the circular buffer to a WAV file. It handles the case where the desired portion wraps around to the beginning of the buffer. 
This program uses 16-bit integer samples, as specified in your question.
Please note that you'll need to link against PortAudio and libsndfile when compiling this C++ code. This can be done by adding `-lportaudio -lsndfile` to your `g++` command.
This is a basic example and doesn't include error checking or any advanced features. You'll likely need to modify it to suit your specific needs.
*/